optuna/storages/in_memory.py

```diff
@@ -196,174 +229,241 @@ def set_trial_state(self, trial_id, state):
+                self.check_trial_is_updatable(trial_id, trial.state)
 
-            # Check param distribution compatibility with previous trial(s).
-            if param_name in self.param_distribution:
-                distributions.check_distribution_compatibility(
-                    self.param_distribution[param_name], distribution
-                )
+                study_id = self._trial_id_to_study_id_and_number[trial_id][0]
+                # Check param distribution compatibility with previous trial(s).
+                if param_name in self._studies[study_id].param_distribution:
+                    distributions.check_distribution_compatibility(
+                        self._studies[study_id].param_distribution[param_name], distribution
+                    )
 
-            # Check param has not been set; otherwise, return False.
-            if param_name in self.trials[trial_id].params:
-                return False
+                # Check param has not been set; otherwise, return False.
+                if param_name in trial.params:
+                    return False
 
-            # Set param distribution.
-            self.param_distribution[param_name] = distribution
+                # Set param distribution.
+                self._studies[study_id].param_distribution[param_name] = distribution
 
-            # Set param.
-            trial = copy.copy(self.trials[trial_id])
-            trial.params = copy.copy(trial.params)
-            trial.params[param_name] = distribution.to_external_repr(param_value_internal)
-            trial.distributions = copy.copy(trial.distributions)
-            trial.distributions[param_name] = distribution
-            self.trials[trial_id] = trial
+                # Set param.
+                trial = copy.copy(trial)
+                trial.params = copy.copy(trial.params)
+                trial.params[param_name] = distribution.to_external_repr(param_value_internal)
+                trial.distributions = copy.copy(trial.distributions)
+                trial.distributions[param_name] = distribution
+                self._set_trial(trial_id, trial)
 
-            return True
+                return True
 
     def get_trial_number_from_id(self, trial_id):
         # type: (int) -> int
 
-        return trial_id
+        with self._lock:
+            self._check_trial_id(trial_id)
+
+            return self._trial_id_to_study_id_and_number[trial_id][1]
 
     def get_best_trial(self, study_id):
         # type: (int) -> FrozenTrial
 
-        if self.best_trial_id is None:
-            raise ValueError("No trials are completed yet.")
-        return self.get_trial(self.best_trial_id)
+        with self._lock:
+            self._check_study_id(study_id)
+
+            best_trial_id = self._studies[study_id].best_trial_id
+            if best_trial_id is None:
+                raise ValueError("No trials are completed yet.")
+            return self._get_trial(best_trial_id)
```

Comment:
Should this be a deep copy if we are to keep the same behavior?

Reply:
Yes, this should be self.get_trial(best_trial_id). I'll fix this.